srtp_writer_future.go

```diff
@@ -0,0 +1,95 @@
+// +build !js
+
+package webrtc
+
+import (
+	"io"
+	"sync/atomic"
+
+	"github.com/pion/rtp"
+	"github.com/pion/srtp"
+)
+
+// srtpWriterFuture blocks Read/Write calls until
+// the SRTP Session is available
+type srtpWriterFuture struct {
+	rtpSender      *RTPSender
+	rtcpReadStream atomic.Value // *srtp.ReadStreamSRTCP
+	rtpWriteStream atomic.Value // *srtp.WriteStreamSRTP
+}
+
+func (s *srtpWriterFuture) init() error {
+	select {
+	case <-s.rtpSender.stopCalled:
+		return io.ErrClosedPipe
+	case <-s.rtpSender.transport.srtpReady:
+	}
+
+	srtcpSession, err := s.rtpSender.transport.getSRTCPSession()
+	if err != nil {
+		return err
+	}
+
+	rtcpReadStream, err := srtcpSession.OpenReadStream(uint32(s.rtpSender.ssrc))
+	if err != nil {
+		return err
+	}
+
+	srtpSession, err := s.rtpSender.transport.getSRTPSession()
+	if err != nil {
+		return err
+	}
+
+	rtpWriteStream, err := srtpSession.OpenWriteStream()
+	if err != nil {
+		return err
+	}
+
+	s.rtcpReadStream.Store(rtcpReadStream)
+	s.rtpWriteStream.Store(rtpWriteStream)
+	return nil
+}
+
+func (s *srtpWriterFuture) Close() error {
+	if value := s.rtcpReadStream.Load(); value != nil {
+		return value.(*srtp.ReadStreamSRTCP).Close()
+	}
+
+	return nil
+}
+
+func (s *srtpWriterFuture) Read(b []byte) (n int, err error) {
```

Comment:
What if we rewrote this method (and similar ones below) to:

```go
func (s *srtpWriterFuture) Read(b []byte) (n int, err error) {
	value := s.rtcpReadStream.Load()

	if value == nil {
		if err := s.init(); err != nil {
			return 0, err
		}

		value = s.rtcpReadStream.Load()
	}

	return value.(*srtp.ReadStreamSRTCP).Read(b)
}
```

Your code looks neater than my suggestion, but it I'm afraid it might be easy to inadvertently trigger infinite recursions by modifying the implementation of `init()` without thinking about the consequences it might have in these methods (for example if another branch is added that does not set `rtcpReadStream`, but does not return an error).

Reply:
Oh yea, much better to be defensive. I added a `nil` check to guard against that.